src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts

```diff
@@ -0,0 +1,352 @@
+    };
+};
+
+const createNode = (input: TextEditorNode | string, schema: Schema): Node => {
+    if (typeof input === 'string') {
+        return schema.text(input);
+    }
+
+    const node = input.node;
+
+    if (typeof node === 'string') {
+        return schema.text(node);
+    }
+
+    const customNode = getCustomNode(node.tagName, schema);
+
+    const childNodes: Node[] = (input.children ?? [])
+        .map((child) => createNode(child, schema))
+        .filter(Boolean);
+
+    return customNode.create(node.attributes, childNodes);
+};
+
+const getCustomNode = (name: string, schema: Schema): NodeType => {
+    const customNode = Object.values(schema.nodes).find(
+        (prosemirrorNode) => prosemirrorNode.name === name,
+    );
+
+    if (!customNode) {
+        throw new Error(
+            `A custom element hasn't been registered for node ${name}`,
+        );
+    }
+
+    return customNode;
+};
+
+const getTriggerStartPosition = (view: EditorView): number => {
+    return view.state?.selection?.$from?.pos;
+};
+
+const isTrigger = (
+    key: string,
+    validTriggers: TriggerCharacter[],
+): key is TriggerCharacter => {
+    return key.length === 1 && validTriggers.includes(key as TriggerCharacter);
+};
+
+const shouldTrigger = (state: EditorState): boolean => {
+    const { $from } = state.selection;
+
+    if ($from.pos === 1) {
+        return true;
+    }
+
+    // Getting the position immediately before the current selection
+    const prevPos = $from.pos - 1;
+
+    if (prevPos > 0) {
+        // allow trigger if the cursor is at the start of a new paragraph
+        if ($from.parentOffset === 0) {
+            return true;
+        }
+
+        const prevChar = state.doc.textBetween(prevPos, $from.pos);
+
+        return prevChar === ' ' || prevChar === '\n';
+    }
+
+    return false;
+};
+
+const stillHasTrigger = (
+    state: EditorState,
+    activeTrigger: string,
+    triggerPosition: number,
+    triggerLength: number,
+): boolean => {
+    const cursorPosition = state.selection.$from.pos;
+
+    if (
+        cursorPosition < triggerPosition ||
+        cursorPosition > triggerPosition + triggerLength + 1
+    ) {
+        return false;
+    }
+
+    return (
+        state.doc.textBetween(triggerPosition, triggerPosition + 1) ===
+        activeTrigger
+    );
+};
+
+const getTriggerEventDetail = (
+    view: EditorView,
+    trigger: TriggerCharacter,
+    value: string,
+): TriggerEventDetail => {
+    return {
+        trigger: trigger,
+        textEditor: inserterFactory(view),
+        value: value,
+    };
+};
+
+const sendTriggerEvent = (
+    type: 'triggerStart' | 'triggerStop' | 'triggerChange',
+    view: EditorView,
+    trigger: TriggerCharacter,
+    value: string,
+) => {
+    const event = new CustomEvent<TriggerEventDetail>(type, {
+        detail: getTriggerEventDetail(view, trigger, value),
+        bubbles: true,
+        composed: true,
+    });
+    view.dom.dispatchEvent(event);
+};
```

Comment:
This might be fine, but I think it would be better if we emit the events from the component instead. We can have a callback on the component that we invoke from here that emits the events, e.g.

```ts
class TextEditor {
    @Event()
    public triggerStart: EventEmitter<TriggerEventDetail>;

    private emitTriggerStart(data) {
        this.triggerStart.emit(...);
    }
}
```